### ISSUE TITLE
scoping out a rust generator

### ISSUE THREAD

**Issue by sneakers-the-rat** [linkml/linkml#2360]
**What is your feature request?**
Id like to make a rust generator, because eventually I will need to write schemaview in rust to get perf in line with p2p applications and swapping and mutating schemas back and forth, and if we have rustgen then we can generate the metamodel and return fully induced classes via pyO3. That also should let us compile schemaview to wasm and directly use linkml in the browser instead of via pyodide.

Making this issue to a) see what we would want as an MVP for rustgen, and b) see if there are any other ppl interested. I think we dont need a full implementation of the metamodel, since the metamodel is comparatively simple vs. What can be expressed with it, as long as we can express and induce classes with it then that's plenty for my purposes.

Initial sketching
- [x] basics: generating structs with fields and types - definitely possible.
- [ ] slots? metadata? Can we do something like Python's annotated types to make slots with metadata?
- [ ] pyO3: optional pyo3 generation
- [ ] dynamic compilation? Is this possible and reasonable? We would need it for testing
- [ ] union types? I think you need to make an enum for this but they dont quite behave like python unions for maybe obvious reasons
- [ ] vec and hashmap: what are the limitations vs. python lists and dicts?
- [ ] can we generate classes that mimick dataclasses?
- [ ] constraints on fields: cardinality, regex patterns, etc.
- [ ] ..?

**How important is this feature to you?** Select from the options below:
• Medium - can do work without it; but it's important (e.g. to save time or for convenience).

**Comment by cmungall** [comment]
Some of this may be useful but it looks like you are implicitly following this:
https://linkml.io/linkml/howtos/port-linkml.html

**Comment by cmungall** [comment]
Putting high level design issues here rather than the PR, which is liable to get in the weeds:

- what is the high level mapping? Do classes, enums map to structs, traits, enums, impls?

**Comment by sneakers-the-rat** [comment]
ya ya ya - 

* Classes -> Structs ([generation](https://github.com/sneakers-the-rat/linkml/blob/91d513d380d0ea35c64c521d49e1a33d2b02e41c/linkml/generators/rustgen/rustgen.py#L254-L263), [template](https://github.com/sneakers-the-rat/linkml/blob/91d513d380d0ea35c64c521d49e1a33d2b02e41c/linkml/generators/rustgen/templates/struct.rs.jinja))
* Enums -> Enums ([generation](https://github.com/sneakers-the-rat/linkml/blob/91d513d380d0ea35c64c521d49e1a33d2b02e41c/linkml/generators/rustgen/rustgen.py#L211-L219), [template](https://github.com/sneakers-the-rat/linkml/blob/91d513d380d0ea35c64c521d49e1a33d2b02e41c/linkml/generators/rustgen/templates/enum.rs.jinja))
* Slots -> Types  - ([generation](https://github.com/sneakers-the-rat/linkml/blob/91d513d380d0ea35c64c521d49e1a33d2b02e41c/linkml/generators/rustgen/rustgen.py#L228-L238), [template](https://github.com/sneakers-the-rat/linkml/blob/91d513d380d0ea35c64c521d49e1a33d2b02e41c/linkml/generators/rustgen/templates/typealias.rs.jinja)) - currently at least. That's just treating slots as type-like things. I think we can add metadata to types using attribute-like macros that expand into trait impls. That sort of means that metamodel types need to be both types and traits, which could be a fun meta-meta-programming problem or could also send us all straight to hell
* Types -> Types  ([generation](https://github.com/sneakers-the-rat/linkml/blob/91d513d380d0ea35c64c521d49e1a33d2b02e41c/linkml/generators/rustgen/rustgen.py#L198-L204), [template](https://github.com/sneakers-the-rat/linkml/blob/91d513d380d0ea35c64c521d49e1a33d2b02e41c/linkml/generators/rustgen/templates/typealias.rs.jinja)) - same caveats

There are probably more elegant and sophisticated mappings to explore, e.g. subsets could be traits, etc.

The stuff that i'm not so sure about and i think we'll have to write a crate with a bunch of macros in it is stuff like
- boolean/etc. logic: i think that's the way this is generically called in linkml, i am referring to using `any_of`, `one_of`, etc. which i most commonly see in ranges, so my internal language for that is "type unions" or "type intersections" and so on. I don't think there is an elegant way of saying "something can be this or that" in rust
- attaching linkml schema metadata: i think we will need to write macros for this, i think usually these have to be accessor methods and there isn't a generic way to, say, define a `uri` field for a slot represented as a type.
- class ranges as inheritance, or really anything inheritance - e.g. if someone specifies a `ClassA` as a range and that should normally allow all of its subclasses, how we would do something like that. So that makes me think that we need to write macros for every attribute, slot, and class so a subclass could declare that it's implementing the methods for a given class, but i haven't gotten there yet.

**Comment by cmungall** [comment]
Continuing part of email discussion with @eilis

This is what a target for conversion of the obographs schema might look like
https://github.com/ielis/obographs/blob/master/src/model.rs

happily this comports with your design above!

What are your thoughts on how attributes and traits can be incorporated? I would assume that serde for serialization and deserialization are pretty standard. One thing that held up javagen was being able to handle the myriad combinations of annotations and frameworks (lombok, hibernate, jackson, ...) people might want to compose in.

**Comment by sneakers-the-rat** [comment]
> What are your thoughts on how attributes and traits can be incorporated?
I don't have many yet, but the nonlocality of trait implementations is a definite strength, where you can have the struct autogenerated in one file and then manually write trait impls elsewhere and not have them clobbered (not sure if this is what you were asking). 

> handle the myriad combinations of annotations and frameworks
So far i have been writing in a pretty literal way, not trying to preoptimize generality before we know what we need (and also learning rust at the same time so i don't have a clear picture of what that generality should even look like yet). The template classes and build result classes are same as pydanticgen, built specifically to be able to support nesty build results being able to return dependencies and whatnot to the top level, and currently i have it so that any crate import that has a version attached to it is also added to the generated `Cargo.toml`. I think eventually we can just make "feature" toggles that are a) a set of dependencies/imports, b) a some simple rules like "apply to all slots" or "apply to slots with range x", and c) values to pass to the templates so we can handle whatever wild ass combo of extras people want. encountering this already a bit with pyo3 and serde.

**Comment by sneakers-the-rat** [comment]
Alright I went and updated the root comment with some milestones and a fuller set of tasks.

- The first set is just the minimal set of "absolute basics" that i think would be reasonable to call https://github.com/linkml/linkml/pull/2364 good. That's mostly me setting up the generator in the same structure as pydanticgen, generalizing some of the stuff i wrote for that, and making it possible for other people to contribute
- After that's merged, then we mark the generator as experimental and emit a warning to that effect when people use it, and then each of the tasks there can be worked on independently as small PRs - how about when one of us is working on one of those we raise an issue/pr and assign it, linking back here, and adding to the `rustgen-v1` milestone, so we know what someone is working on and don't double effort.
- Then I made a rough distinction between "the stuff that i have seen people using regularly" and "the stuff that i think is cool and we should support, but is maybe less common" like rules and some of the more complex validation slots. Stuff can be moved around between `v1` and `v2`, but the iea is that `v1` would be when we mark rustgen as available for use, and then `v2` would be continued effort to make it work it should over time.

I've split out any of the work on `linkml-core` from this issue so it can focus on the generator, but ya hopefully that is some scaffolding!

**Comment by sneakers-the-rat** [comment]
Would love to have your contribs since my rust is decidedly inelegant. I have been waylaid by another project but still want to do this longer term

**Comment by kervel** [comment]
ok. i think the inheritance issue is now kindof a decision point. it will determine how flexible we are in the future. Linkml allows to do stuff that we need to think about (eg breaking liskov principle by using slot_usage in subtypes).

btw, i think my goals are about the same as yours, we are also hitting a performance wall. and a re-implementation of schemaview in rust would also help my case.

**Comment by kervel** [comment]
let me launch a proposal with a request for comments then ...

* for every linkml class i both generate a `Data`struct and a trait. Then for child classes i can implement the parent trait
* by default the behaviour of all slots is dynamic (so with dyn containers).
* an annotation can be set in the linkml to make it static (for performance critical code). in this case, a slot will be implemented as having as type a union of all the types known at generation time

i used chatgpt to build an elaborated example:

>  We generate traits as public interface for our structs

```rust
pub trait Vehicle {
    #[inline]
    fn get_mode(&self) -> &str;
}

pub trait Car: Vehicle {
    #[inline]
    fn get_name(&self) -> &str;
}

pub trait Sedan: Car {
    #[inline]
    fn get_seats(&self) -> u8;
}
```

> The data is held in a struct. here i use composition to get the fields of the parent struct, but we could also not do that and simply repeat the fields. Every struct is public but the fields are private. There is a new function to construct, that can also do validation.

```rust
// Structs

pub struct VehicleData {
    mode: String,
}

impl VehicleData {
    #[inline]
    pub fn new(mode: impl Into<String>) -> Self {
        Self { mode: mode.into() }
    }
}

pub struct CarData {
    vehicle: VehicleData,
    name: String,
}

impl CarData {
    #[inline]
    pub fn new(mode: impl Into<String>, name: impl Into<String>) -> Self {
        Self {
            vehicle: VehicleData::new(mode),
            name: name.into(),
        }
    }
}

pub struct SedanData {
    car: CarData,
    seats: u8,
}

impl SedanData {
    #[inline]
    pub fn new(mode: impl Into<String>, name: impl Into<String>, seats: u8) -> Self {
        Self {
            car: CarData::new(mode, name),
            seats,
        }
    }
}

```

> We then implement all relevant traits from the structs. This can become a lot for deep class hierarchies.

```rust

// Trait Implementations

impl Vehicle for CarData {
    #[inline]
    fn get_mode(&self) -> &str {
        self.vehicle.get_mode()
    }
}

impl Vehicle for SedanData {
    #[inline]
    fn get_mode(&self) -> &str {
        self.car.get_mode()
    }
}

impl Car for CarData {
    #[inline]
    fn get_name(&self) -> &str {
        &self.name
    }
}

impl Car for SedanData {
    #[inline]
    fn get_name(&self) -> &str {
        self.car.get_name()
    }
}

impl Sedan for SedanData {
    #[inline]
    fn get_seats(&self) -> u8 {
        self.seats
    }
}
```

> Now we can look at how we use classes as range of a slot. As an example i use multivalued slot, but single value is trivial.

for slots then this would be the default behaviour:

```rust
pub struct DynamicVehicleCollection {
    vehicles: Vec<Box<dyn Vehicle>>,
}
```

for performance critical code this would then avoid dynamic dispatch:

```rust
pub enum VehicleSlotRange {
    Car(CarData),
    Sedan(SedanData),
    Vehicle(VehicleData),
}

pub struct StaticVehicleCollection {
    vehicles: Vec<VehicleSlotRange>,
}
```

My rationale for this:

* This reminds me of the 90s era javabeans :-)
* The Deref thing is cool but doesn't always work (rust deref coercion can work recursively but only in certain cases)
* I think we can't make a decision between static and dynamic here. It will really depend on the usecase, but dynamic is probably a good default. But for instance when implementing a rust schemaview we would probably use a version of the metamodel that is all static.
* It puts the boilerplate in the generated code (all the trait impls) rather than in the manually written code (no need for doing .as_car() or so when we want to convert a sedan to a car)

**Comment by cmungall** [comment]
> ok. i think the inheritance issue is now kindof a decision point. it will determine how flexible we are in the future. Linkml allows to do stuff that we need to think about (eg breaking liskov principle by using slot_usage in subtypes).

non-monotonicity is strongly discouraged, I think you can build this assuming monotonicity

**Comment by sneakers-the-rat** [comment]
I'll be able to think more about this in more detail probably Thursday afternoon, but for now I just wanted to link the (crappy) first pass version I had, generated from the PR linked in OP. We probably want to work on a repo within the linkml org but also happy to give you admin on that or whatever works for prototyping for you.
https://github.com/p2p-ld/linkml-core/blob/main/src/metamodel/src/lib.rs

**Comment by sneakers-the-rat** [comment]
I was starting with the metamodel because we need it to bootstrap the rest, and the metamodel is blessedly simple without any of the fancier rules or polymorphism or unions or whatevs that actual schemas use. Might be a good way to get code running and then be able to build on top of that. I think "make metamodel and schemaview" can be considered different from "make a general purpose rust generator" for the sake of subdividing milestones, because metamodel is a (structurally) simple schema and schemaview is a set of more or less straightforward operations on metamodel classes

**Comment by kervel** [comment]
Hi  @sneakers-the-rat i saw that in your PR you also made a couple of seemingly useful changes in the common template, and these are now conflicting with main. Since the PR will likely live for a while, wouldn't it be a good idea to split those changes into a separate PR and get them merged in ?

also, on your rust schemaview impl: was it in your plan to have the exact same schemaview impl ? because if not we could deviate from it in some useful ways.

**Comment by kervel** [comment]
also, i see that your tests now still fail because of something i didn't know yet:

this is in a linkml schema:

```yaml
prefixes:
  linkml: https://w3id.org/linkml/

...
```

however, a prefix has a range: Prefix (not string) in the metamodel:

```yaml
  prefix:
    rank: 12
    description: prefix URI tuple
    slots:
      - prefix_prefix
      - prefix_reference
    in_subset:
      - SpecificationSubset
      - BasicSubset
```

(where prefix_prefix has key:true) 

there is a `_normalize_inlined_as_dict` function that translates the string to Prefix objects. It does this:

```python
...
       elif not isinstance(v, list):
               order_up(k, slot_type(*[k, v]))
...
```

so it basically passes the key and the value as arguments to the Prefix class, which works because Prefix has two slots one with rank 1 and one with rank 2.

we could do the equivalent in rust `serde` by writing a custom deserializer. but the current form is very flexible in what it accepts. also, i feel like relying on the rank of the attribute is not very robust. 

this normalize_inlined_as_dict is not implemented in other generators afaik, i just checked, pydantic generation of metamodel also fails to load the metamodel as yaml.

**Comment by kervel** [comment]
hi @sneakers-the-rat 

i fixed some things (i made a PR to your repo)

https://github.com/sneakers-the-rat/linkml/pull/9

* fix the issue above with a custom deserializer. now the metamodel "loads".
* change the generation of pyo3 and serde to rust feature gates rather than jinja template ifs. i think if we ever want to create a linkml-core package, we probably want to be able to compile it with or without python support.

however, i also found that we have to do inheritance in order to do something meaningful: the metamodel already uses quite some inheritance and polymorphism (look at AnyValue and friends). which is a bummer because more work :-)

**Comment by kervel** [comment]
small heads up. i made some progress. generated code for a class with subtypes now looks like so:

https://github.com/Kapernikov/rust-linkml-core/blob/new-rustgen/src/metamodel/src/lib.rs#L529

its a lot of code right now (most could be automated with macros but what's the point of that as its generated code ? it will only increase compile times).

it uses serde tags when type designators are present.

the idea is to implement polymorphism in a separate mod in the same crate (that is optional), which will contain traits for each class that will be implemented for all subclasses and by the "OrSubtype" enums.

**Comment by sneakers-the-rat** [comment]
Wow that looks great. You clearly know what you're doing better than I do, so I am gonna give you commit privs to my fork so you dont have to wait on me. I likely won't have time for this for awhile, but am excited to see it

**Comment by kervel** [comment]
another update!
polymorphism for superclasses and mixins:

the generated Extension class (which has two subclasses):
https://github.com/Kapernikov/rust-linkml-core/blob/new-rustgen/src/metamodel/src/poly.rs#L27

the Extensible class has many more subclasses, so way more code.
https://github.com/Kapernikov/rust-linkml-core/blob/new-rustgen/src/metamodel/src/poly.rs#L92

i don't use trait objects here, i do all polymorphism with enums. trait objects would involve copying (i think)

**Comment by kervel** [comment]
now the rust-generated metamodel succesfully loads the whole linkml metamodel including all imports (see new test in io.rs)

i am now at a point where i would need some feedback:

* i avoided trait objects for now and used enums to model a "type or subtype" in collections. the advantage is that the native struct always remains visible. the disadvantage is that now its a lot of (generated) code, and the size of the enum is always as big as the biggest subclass, which might be a problem for large hierarchies. but going for trait objects (`Vec<Box<dyn Element>>`) instead of enums (`Vec<Box<ElementOrSubtype>>`)

* if we were to merge the rust generator into main, how could we meaningfully test it without bringing in rust as a dependency

* lots of the complexity stems from the fact that i have to support both boxed (`Vec<Box<Element>>`) attributes and non-boxed attributes (`Vec<Element>`). especially for the Box this costs a lot: i have to implement `IntoPyObject` for all boxed objects (which means that a python API would not be able to use the functions only the fields of those objects),  because i cannot `#[pyclass]` a box. Also, i cannot implement traits for Box<T> (rust orphan rule). We will face the same trouble if we add wasmbindgen for typescript bindings.

* i had to implement rust variants of what happens in the python classes `__post_init` (all those normalize_ functions) to make it load. I don't think there is a specification of what yaml should be accepted and how these post init functions should behave. or is there ?

**Comment by sneakers-the-rat** [comment]
Very cool. I'll take a look by Friday :):)

> if we were to merge the rust generator into main, how could we meaningfully test it without bringing in rust as a dependency

I personally think this is fine - we would have to be a little bit strategic about out tests to avoid needing to compile a million things, but having optional pythin and system dependencies in ci is normal and done for a number of the other generators. I also think it would probably be good to keep as a separate package - adding more generators into the main package is imo just digging a deeper hole for when we eventually get to  https://github.com/linkml/linkml/issues/1784 

> i had to implement rust variants of what happens in the python classes __post_init (all those normalize_ functions) to make it load

Aw shucks I was hoping we could leave those behind. Is this what you're looking for? https://linkml.io/linkml-model/latest/docs/specification/06mapping/#collection-forms
Not sure how closely that maps on to the impl

**Comment by kervel** [comment]
@sneakers-the-rat thx, didn't know about that. there seems to be one thing missing: multivalued properties also accept single values (a string translates to a list of strings with length 1). i implemented that only for primitive values, not for nested objects. it is used for the "notes" field in the metamodel.

there is a second pr in your repo. i'll try to follow up with some simple python based tests (eg on generating the correct rust type). the real rust tests i would put in the rust linkml-core impl package (for which i also put a pr), does that make sense ?

relevant PRs:

https://github.com/p2p-ld/linkml-core/pull/1
https://github.com/sneakers-the-rat/linkml/pull/10

**Comment by sneakers-the-rat** [comment]
i am gonna make you a maintainer of both of those, you shouldn't have to wait on my permission when you're the one doing all the work!

**Comment by sneakers-the-rat** [comment]
also lmk what i should update in the top checklist :)

**Comment by kervel** [comment]
hello @sneakers-the-rat . it took a while, but i created a new PR with your comments. i prefer the pr route if you don't mind, rather than committing directly, at least i get some feedback that way :-)

in the mean time i have been toying with the rust linkml core impl, created a toy yaml->ttl conversion to test things out.
the code lives here for now https://github.com/Kapernikov/rust-linkml-core/tree/new-rustgen

i think the checkboxes are more or less accurate on top, i didn't look at ifabsent yet at all, probably not that easy.

do you think the ifabsent is really needed for this to be merged in linkml (and mark it as experimental) ? i don't think adding support for it later will cause backwards compat problems.

**Comment by sneakers-the-rat** [comment]
I think it's not a dealbreaker but something we should probably have a plan for. I came up against this challenge too and sorta concluded we would need a macro for it because to my tiny rust brain everything is a problem that can be solved by more macros lol. What about it seems hard to you? There are a few flavors of ifabsent, some of them are just plain static defaults, but some can be expressions which I imagine would be much harder

**Comment by kervel** [comment]
well its about implementing all the flavours, and i don't know if ifabsent is supposed to also work for non-primitive value (can you set an ifabsent on an inlined class ranged slot ?)

i never used ifabsent in my projects so i actually don't know it that well ..

about macros: i didn't do that yet, this i my reasoning: we generate code. macros make code shorter and easier to read at the expense of compile time and complexity. but since the code is generated extra LOCs don't cost anything, so i didn't bother.

so now every time i wanted a macro i just made the rustgen spit out an ordinary function.

**Comment by kervel** [comment]
another thing that will be as difficult as ifabsent: proper support for min_value and max_value. they are defined as range:Anything and Anything is a class ...

```shell
$ ./target/debug/linkml-schema-validate ./src/runtime/tests/data/personinfo.yaml
Error: Error { path: Path { segments: [Map { key: "slots" }] }, 
   original: Error("at `maximum_value`: Invalid type integer `999`. Expected struct Anything", line: 1, column: 1) }
```

```rust
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "pyo3", pyclass(subclass, get_all, set_all))]
pub struct Anything {
}
```

same for AnyValue ...

how would i represent a Anything in rust ?

**update** i committed a special "anything" class now. it just wraps the serde_value::Value for now. which is not great because won't work if no serde dependency. but i really don't know of any other option.

**Comment by sneakers-the-rat** [comment]
Huh that seems like a problem in the metamodel- when would a min/max value not be a float or int?

**Comment by kervel** [comment]
did some more work:

* use merge crate to merge two objects of the same class (for instance in rust i can now merge two slotdefinitions to implement slot_usage)
* support anyvalue

the toy rust linkml runtime https://github.com/Kapernikov/rust-linkml-core now has some features too:
* more work on SchemaView
* added ClassView and SlotView
* all class names and slot names are parsed and treated as identifiers (name or uri or curie). which kindof fixes the [namespaces](https://github.com/orgs/linkml/discussions/1739) problem without any changes in the metamodel.
* a working linkml yaml to ttl convertor (i need to test this with a big dataset) as an example.

for instance (simplified real world example i am facing):

```yaml
id: http://example.com/geo
name: geo
prefixes:
  geo: http://example.com/geo/
  linkml: https://w3id.org/linkml/
imports:
  - linkml:types
default_prefix: geo
classes:
  Geometry:
    description: Base geometry class
```

```yaml
id: http://example.com/era
name: era
prefixes:
  era: http://example.com/era/
  geo: http://example.com/geo/
  linkml: https://w3id.org/linkml/
imports:
  - linkml:types
  - http://example.com/geo
default_prefix: era
classes:
  Geometry:
    is_a: geo:Geometry
    description: Extended geometry
```

gets parsed correctly (you only need to use a uri/curie when there would be ambiguity, using a class name still works)

i have to add an annotation to the metamodel:

```yaml
  slot_definition:
    rank: 3
    description: an element that describes how instances are related to other instances
    annotations:
      "rust.linkml.io/generate/merge":
        value: true
```

to trigger generation of the "merge" function, needed to create induced_slots

**Comment by kervel** [comment]
some progress (in the PR to @sneakers-the-rat repo):

* functions that return simple types (types implementing Copy) now have a simpler signature (they return for instance an `Option<bool>` instead of an `Option<&bool>`)
* for all multivalued containers (vec, hashmap) i so far didn't use Option for optional things (my original reasoning: when its not in the yaml, just make an empty list). however, an empty list is not the same as a missing value. so now its changed so i have Option<Vec> and Option<HashMap> when stuff is optional. which is a bit less ergonomic, but i think more correct.

**Comment by clbarnes** [comment]
On inheritance, I think that the enum solution is not ideal because you have to list all subtypes, which means nobody can extend your schema with their own subclasses, they'd have to create a whole new crate implementing the entire class tree.

I think that the most rusty solution is to create a type with private fields including some sort of `_super: ParentType` and `_mixins: (Mixin1, Mixin2, ...)`, a trait with getters and setters for all fields, and possibly a builder for each as well. The implementations of the trait on children just delegate to the relevant fields; the builder methods would also delegate to superclass builders. The accessor trait for a child could contain copies of the parent's methods or the child trait could use the parent's accessor trait as a trait bound and just implement each of the superclasses' accessors (may be more confusing for the end user's imports, because it wouldn't necessarily be clear which superclass controlled which field).

Important features would be for users to be able to define their own builders and accessor traits and hide the custom ones (which rust's privacy can handle), because of cases where particular fields need to be kept in sync in ways the spec itself can't express.

I guess type-or-subtype fields would need to be handled by generics, which is not ideal as the chain of generics could be very long. On ser/de, I suppose you have to know what type you're getting out at the other end anyway and any extra fields can be ignored.

**Comment by kervel** [comment]
hi @clbarnes i missed your comment (maybe we should close the old issue and start a new one ?)

to make the traits (i generate in poly.rs) truly "open world" one would need to stop assuming things about ownership, and about implementations. Combine that with mutations and all object-valued trait objects would then have to return stuff like `Arc<RwLock<dyn>>` and do a lot of copying / management. it would start to feel like a runtime quickly.

I also think treating the structs as data rather than behaviour has real advantages: the fields are exactly where you'd expect them, you can read them directly without navigating _super / _mixins to find a value, and you don't need the traits at all if you don't want them.

the current implementation is a trade off, and for me i took the trade off as "i want something low overhead, that was for me the first reason to not stick with python". and if  i want to create subclasses i'd probably do that in a new linkml derived schema rather than rust code. 

But if we can find a way to make the traits more "open world" and not leak as much implementation in them that would be nice. I tried hard here and i never found a fully clean way to do it.
